great_tables/_modify_rows.py

```diff
@@ -178,3 +191,175 @@ def with_id(self: GTSelf, id: str | None = None) -> GTSelf:
+        can facilitate method chaining.
+
+    Examples
+    --------
+    Let's use a subset of the `sp500` dataset to create a table with grand summary rows. We'll
+    calculate min, max, and mean values for the numeric columns. Notice the different
+    approaches to selecting columns to apply the aggregations to: we can use polars selectors
+    or select the columns directly.
+
+    ```{python}
+    import polars as pl
+    import polars.selectors as cs
+    from great_tables import GT, vals, style, loc
+    from great_tables.data import sp500
+
+    sp500_mini = (
+        pl.from_pandas(sp500)
+        .slice(0, 7)
+        .drop(["volume", "adj_close"])
+    )
+
+    (
+        GT(sp500_mini, rowname_col="date")
+        .grand_summary_rows(
+            fns={
+                "Minimum": pl.min("open", "high", "low", "close"),
+                "Maximum": pl.col("open", "high", "low", "close").max(),
+                "Average": cs.numeric().mean(),
+            },
+            fmt=vals.fmt_currency,
+        )
+        .tab_style(
+            style=[
+                style.text(color="crimson"),
+                style.fill(color="lightgray"),
+            ],
+            locations=loc.grand_summary(),
+        )
+    )
+    ```
+
+    We can also use custom callable functions to create more complex summary calculations.
+    Notice here that grand summary rows can be placed at the top of the table and formatted
+    with currency notation, by passing a formatter from the `vals.fmt_*` class of functions.
+
+    ```{python}
+    from great_tables import GT, style, loc, vals
+    from great_tables.data import gtcars
+
+    def pd_median(df):
+        return df.median(numeric_only=True)
+
+
+    (
+        GT(
+            gtcars[["mfr", "model", "hp", "trq", "mpg_c"]].head(6),
+            rowname_col="model",
+        )
+        .fmt_integer(columns=["hp", "trq", "mpg_c"])
+        .grand_summary_rows(
+            fns={
+                "Min": lambda df: df.min(numeric_only=True),
+                "Max": lambda df: df.max(numeric_only=True),
+                "Median": pd_median,
+            },
+            side="top",
+            fmt=vals.fmt_integer,
+        )
+        .tab_style(
+            style=[style.text(color="crimson", weight="bold"), style.fill(color="lightgray")],
+            locations=loc.grand_summary_stub(),
+        )
+    )
+    ```
+
+    """
+    if columns is not None:
+        raise NotImplementedError(
+            "Currently, grand_summary_rows() does not support column selection."
+        )
+
+    # summary_col_names = resolve_cols_c(data=self, expr=columns)
+
+    for label, fn in fns.items():
+        row_values_dict = _calculate_summary_row(self, fn, fmt, missing_text)
+
+        summary_row_info = SummaryRowInfo(
+            id=label,
+            label=label,
+            values=row_values_dict,
+            side=side,
+        )
+
+        self._summary_rows_grand.add_summary_row(summary_row_info)
```

Comment:
Let's ensure we don't mutate here and ensure summary row data objects are not allowed to mutate themselves